src/watchful/client2.py

```diff
@@ -314,23 +315,103 @@ def set_hub_url(self, url) -> typing.Dict[str, str]:
     # The functions below interface with a "hub" instance, and thus
     # require authentication.
 
-    def login(self, email: str, password: str) -> None:
+    def login(
+        self, email: str, password: str
+    ) -> typing.Union[str, typing.Dict[str, str]]:
         """Log in to hub."""
-
-    def publish(self, token: str) -> None:
+        credentials = base64.b64encode(f"{email}:{password}".encode()).decode(
+            "utf-8"
+        )
+        # Remote functions return a summary. In this case, we'll ignore the
+        # summary, as it's mostly irrelevant for what we need.
+        response = self._session.post(
+            urljoin(self._root_url, "remote"),
+            json={"verb": "login"},
+            headers={"Authorization": f"Basic {credentials}"},
+            timeout=self.timeout,
+        )
+        try:
+            return response.json()
+        except ValueError:
+            return response.text
+
+    # XXX: rockstar (22 May 2023) - The following functions all required a
+    # token. It's likely that we should just capture the remote token somewhere
+    # and re-use it as part of the session.
+    def publish(self, token: str) -> typing.Union[str, typing.Dict[str, str]]:
         """Publish to hub."""
+        response = self._session.post(
+            urljoin(self._root_url, "remote"),
```

Comment:
Each of these methods has a common block in the `self._session.post(...)` call. Should these be extracted into a separate method that accepts the `verb` as the argument?